### PR TITLE
Add testing prior minor to our cross-version tests

### PR DESCRIPTION
**Pull request by vladsud**
Ensure that we at least test (prior minor, current version) mix (both directions) when running cross-version tests, i.e. ensure that document format (ops, summaries) is compatible across two closest minor releases.

This change by itself does not add that much coverage but creates possible framework for future changes.
Current tests only validate that with a mix of 1.x and latest 2.x. This is good, but we might miss cases that way. Also, tests might be disabling cross-compat tests altogether because they are testing recently added features - we would love to move toward a direction where such tests only disable parts of cross-version tests - the part that has 1.x in it but enable the other part - testing N / N -1. We can even introduce a new mode in addition to "Full Compat" that enables N / N-1 in next release, such that developers can add tests and not need to go back to enable partial compat mode (all TBD, and not prescribed here).

It will increase time to run tests, as it doubles number of cross-version pairs. That said, it adds only 2 configs to already 13 configs we are testing in Full compat mode, and not all tests use Full compat mode.
I've tested it on my box, running only end-to-end tests:
1. Without this change: 2m 55 seconds
2. With this change: 3m 18 seconds

So that's 13% increase (based on one sample)
Relative %% will be lower if considering running all the tests (as there are more tests that are not affected by this change).

### CODE REVIEW

File: packages/test/test-version-utils/src/compatConfig.ts [hunk 1]
```diff
@@ -259,6 +259,18 @@ export function isCompatVersionBelowMinVersion(minVersion: string, config: Compa
 	return semver.compare(compatVersion, minReqVersion) < 0;
 }
 
+function genCompatConfig(createVersion: string, loadVersion: string): CompatConfig {
```
Comment (scottn12) [inline packages/test/test-version-utils/src/compatConfig.ts:262]: nit: Would be nice to add a comment describing this function so we know its a helper for `genCrossVersionCompatConfig`

File: packages/test/test-version-utils/src/compatConfig.ts [hunk 2]
```diff
@@ -271,41 +283,30 @@ export function isCompatVersionBelowMinVersion(minVersion: string, config: Compa
...
-					};
-				}),
-			)
-			.reduce((a, b) => a.concat(b))
-			// Filter to ensure we don't create/load with the same version.
-			.filter((config) => config.compatVersion !== config.loadVersion)
-	);
+	const currentVersion = getRequestedVersion(pkgVersion, 0);
+
+	//
```
Comment (scottn12) [inline packages/test/test-version-utils/src/compatConfig.ts:288]: nit: looks like you have extra empty commented lines here?

File: packages/test/test-version-utils/src/compatConfig.ts [hunk 3]
```diff
@@ -271,41 +283,30 @@ export function isCompatVersionBelowMinVersion(minVersion: string, config: Compa
...
+	// Build a list of all the versions we want to test, except current version.
+	//
+	const allDefaultDeltaVersions = defaultCompatVersions.currentVersionDeltas
+		.filter((delta) => delta !== 0) // skip current build
+		.map((delta) => getRequestedVersion(pkgVersion, delta));
+	allDefaultDeltaVersions.push(...defaultCompatVersions.ltsVersions);
+
+	//
```
Comment (scottn12) [inline packages/test/test-version-utils/src/compatConfig.ts:296]: nit: same as above

**Review by scottn12 — APPROVED** [review]
LGTM, left a couple nits as well

**Comment by vladsud** [comment]
FYI - Navin suggested that our pipelines are already slow and there is a desire to make them faster. So maybe this change needs to wait until we have a plan here, not sure. So, time till checkin (in any form) might be unknown - will continue to work with Navin to create clarity here.

**Comment by vladsud** [comment]
That said, it's really interesting to see some tests failing. For one, they do not fail on my local machine (at least local tests, did not try t9s) :)
But also, some of them fail with fluid:telemetry:Container:SubmitMessageWithNoConnection, which should not happen:
It was fixed via https://github.com/microsoft/FluidFramework/pull/20111, and https://github.com/microsoft/FluidFramework/pull/20109explicitly would disable compression in RC2 not to hit it.

Others fail with timeouts, which is a bit harder to diagnose.
Worth looking into, there could be more bugs to be found :)

**Review by agarwal-navin — APPROVED** [review]

File: packages/test/test-version-utils/src/compatUtils.ts [hunk 4]
```diff
@@ -334,7 +333,14 @@ export async function getCompatVersionedTestObjectProviderFromApis(
...
+		// First, check if any of the versions is current version of the package.
+		// Current versions show up in the form of "2.0.0-dev-rc.3.0.0.251800", and semver.compare()
+		// incorrectly compares them with prior minors, like "2.0.0-rc.2.0.1"
+		versionForLoading === pkgVersion
+			? versionForCreating
+			: versionForCreating === pkgVersion
+			? versionForLoading
+			: semver.compare(versionForCreating, versionForLoading) < 0
 			? versionForCreating
 			: versionForLoading;
```
[the remark this answers is not in the document]
Comment (tylerbutler) [reply packages/test/test-version-utils/src/compatUtils.ts:345]: We have a bunch of utility functions in version-utils (part of build-tools) that are intended to make working with our versions a bit easier, but I don't think any of them apply here. Can you explain what you _want_ to do? What versions are you trying to compare? Are you trying to figure out what the in-repo version that corresponds with a build version of `2.0.0-dev-rc.3.0.0.251800`? Or are you expecting all versions produced by our build system to be universally sortable?

Regarding version schema, it's unfortunately not a trivial problem to solve as far as I can tell. The prerelease rules for semver are as follows (emphasis added):

> Precedence for two pre-release versions with the same major, minor, and patch version MUST be determined by comparing each dot separated identifier from left to right until a difference is found as follows:
>
> Identifiers consisting of only digits are compared numerically.
>
> Identifiers with letters or hyphens are compared lexically in ASCII sort order.
>
> Numeric identifiers always have lower precedence than non-numeric identifiers.
>
> **A larger set of pre-release fields has a higher precedence than a smaller set, if all of the preceding identifiers are equal.**

I think it all becomes moot once we are using non-prerelease semvers (i.e. 2.0, 2.1, 3.0, etc.), though.

**Pull request merged**